File: src/source/Yaaf.Database/AbstractApplicationIdentityDbContext.cs
// ----------------------------------------------------------------------------
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.
// ----------------------------------------------------------------------------
using Microsoft.AspNet.Identity.EntityFramework;
using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Data.Entity.Core.Objects;
using System.Data.Entity.Infrastructure;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Yaaf.Database {
	public abstract class AbstractApplicationIdentityDbContext<TUser> : IdentityDbContext<TUser> where TUser : IdentityUser {
		static AbstractApplicationIdentityDbContext ()
		{
			if (String.IsNullOrWhiteSpace (AppDomain.CurrentDomain.GetData ("DataDirectory") as string)) {
				System.AppDomain.CurrentDomain.SetData (
					"DataDirectory",
					System.AppDomain.CurrentDomain.BaseDirectory);
			}
		}

		protected virtual void Init ()
		{
		}

		private void MyInit ()
		{
			Init ();
			Database.Initialize (false);
		}

		public AbstractApplicationIdentityDbContext ()
			: base ()
		{
			MyInit ();
		}

		public AbstractApplicationIdentityDbContext (string nameOrConnection)
			: base (nameOrConnection)
		{
			MyInit ();
		}

		public Task MySaveChanges ()
		{
			return AbstractApplicationDbContext.MySaveChanges (this);
		}
	}
}

File: src/source/Yaaf.Database/AbstractApplicationIdentityDbContext.fs
// ----------------------------------------------------------------------------
// This file is subject to the terms and conditions defined in
// file 'LICENSE.txt', which is part of this source code package.
// ----------------------------------------------------------------------------
namespace Yaaf.Database
open Microsoft.AspNet.Identity.EntityFramework
open System
open System.Collections.Generic
open System.Data.Entity
open System.Data.Entity.Migrations
open System.Data.Entity.Core.Objects
open System.Data.Entity.Infrastructure
open System.Linq
open System.Text
open System.Threading
open System.Threading.Tasks
open Yaaf.Helper

[<AbstractClass>]
type AbstractApplicationIdentityDbContext<'TUser when 'TUser :> IdentityUser >(nameOrConnectionString) as x =
    inherit IdentityDbContext<'TUser>(nameOrConnectionString : string)           
    static do
        if (String.IsNullOrWhiteSpace (AppDomain.CurrentDomain.GetData ("DataDirectory") :?> string)) then
            System.AppDomain.CurrentDomain.SetData (
                "DataDirectory",
                System.AppDomain.CurrentDomain.BaseDirectory)
    interface IUpgradeDatabaseProvider with
      member x.GetMigrator() = x.GetMigrator()
      
      member x.FixScript s = x.FixScript s
      
    abstract FixScript : string -> string
    default x.FixScript s = s

    abstract GetMigrator : unit -> DbMigrator
    default x.GetMigrator () = DatabaseUpgrade.notImpl ()

    member x.MySaveChanges () =
        AbstractApplicationDbContext.MySaveChanges (x)    
